File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode trav = head;
        HashMap<ListNode,Integer> hash_map = new HashMap<ListNode,Integer>();
        if(head == null || head.next == null)
        {
            return false;
        }
        hash_map.put(trav,trav.val);
        trav = trav.next;
        while(hash_map.containsKey(trav) != true)
        {
            hash_map.put(trav,trav.val);
            trav = trav.next;
            if(trav == null)
                break;
        }
        if(trav != null)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        f, s = head, head
        while f and f.next:
            if f.next:
                f = f.next.next
                s = s.next
                if f == s: return True
        return False
            
        